timemachine/cpp/src/local_md_utils.hpp

```diff
@@ -0,0 +1,150 @@
+        nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<double>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else {
+        throw std::runtime_error("unable to cast potential to NonbondedAllPairs");
+    }
+}
+
+void set_nonbonded_ixn_potential_idxs(
+    std::shared_ptr<Potential> pot,
+    const int num_col_idxs,
+    const int num_row_idxs,
+    unsigned int *d_col_idxs,
+    unsigned int *d_row_idxs,
+    const cudaStream_t stream) {
+    if (std::shared_ptr<NonbondedInteractionGroup<float>> nb_pot =
+            std::dynamic_pointer_cast<NonbondedInteractionGroup<float>>(pot);
+        nb_pot) {
+        nb_pot->set_atom_idxs_device(num_col_idxs, num_row_idxs, d_col_idxs, d_row_idxs, stream);
+    } else if (std::shared_ptr<NonbondedInteractionGroup<double>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedInteractionGroup<double>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_col_idxs, num_row_idxs, d_col_idxs, d_row_idxs, stream);
+    } else {
+        throw std::runtime_error("Unable to cast potential to NonbondedInteractionGroup");
+    }
+}
+
+std::shared_ptr<BoundPotential>
+construct_ixn_group_potential(const int N, std::shared_ptr<Potential> pot, const int P, const double *d_params) {
+    std::vector<double> h_params(P);
+    gpuErrchk(cudaMemcpy(&h_params[0], d_params, P * sizeof(*d_params), cudaMemcpyDeviceToHost));
+    std::vector<int> dummy_idxs{0};
+    std::vector<int> shape{P};
+
+    if (std::shared_ptr<NonbondedAllPairs<float>> nb_pot = std::dynamic_pointer_cast<NonbondedAllPairs<float>>(pot);
+        nb_pot) {
+        std::shared_ptr<Potential> ixn_group(
+            new NonbondedInteractionGroup<float>(N, dummy_idxs, nb_pot->get_beta(), nb_pot->get_cutoff()));
+        return std::shared_ptr<BoundPotential>(new BoundPotential(ixn_group, shape, &h_params[0]));
+    } else if (std::shared_ptr<NonbondedAllPairs<double>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double>>(pot);
+               nb_pot) {
+        std::shared_ptr<Potential> ixn_group(
+            new NonbondedInteractionGroup<double>(N, dummy_idxs, nb_pot->get_beta(), nb_pot->get_cutoff()));
+        return std::shared_ptr<BoundPotential>(new BoundPotential(ixn_group, shape, &h_params[0]));
+    } else {
+        throw std::runtime_error("unable to cast potential to NonbondedAllPairs");
+    }
+}
+
+// Recursively flatten the potentials. Important to find specific NonbondedAllPairs potentials for multiple_steps_local which
+// can be wrapped in FanoutSummedPotential or SummedPotential objects. Creates copies of bound potentials, for simplicity and safety.
+void flatten_potentials(
```

Comment:
This makes copies of all of the nested potentials, which is a lot of overhead.

Likely need to add constructors for bound potential to take in gpu memory

Reply:
I'm not sure I understand the necessity of creating copies of the nested potentials. Does this have to do with the fact that local MD modifies the neighborlist used by `NonbondedAllPairs`? Would it be possible to at least avoid copying the other potentials here?

Reply:
This is primarily due to the simplicity of this approach. Had played around with several other versions that avoided copying, but seemed overly complex for a first pass that is likely to change.

The goals are to be able determine the following:
* Find `NonbondedAllPairs` Potentials
* Determine that is only one `NonbondedAllPairs`, hard to verify the case where you have multiple
* Have the parameters to construct the `NonbondedInteractionGroup`

Will try modifying it to return a list of NonBondedAllPairs wrapped as BoundPotentials and only reconstruct SummedPotentials/NonbondedAllPairs.

Reply:
Changed as the approach of flattening only the nonbonded all pairs is simple and gets more directly at the problem.

Does add a bit more nuance in that the BoundPotential of the NonbondedAllPairs can't be modified, but only the underlying potential is modified which is shared with the potential in the FanoutSummedPotential.

Reply:
Done in [5b41c5d](https://github.com/proteneer/timemachine/pull/975/commits/5b41c5d7d354686cf213b781ecd8a7cccf8b912e)